## Part 3: Building a Recommender System with Implicit Feedback

In this tutorial, we will build an implicit feedback recommender system using the [implicit](https://github.com/benfred/implicit) package.

What is implicit feedback, exactly? Let's revisit collaborative filtering. In [Part 1](https://github.com/topspinj/recommender-tutorial/blob/master/part-1-item-item-recommender.ipynb), we learned that [collaborative filtering](https://en.wikipedia.org/wiki/Collaborative_filtering) is based on the assumption that `similar users like similar things`. The user-item matrix, or "utility matrix", is the foundation of collaborative filtering. In the utility matrix, rows represent users and columns represent items. 

<img src="images/utility-matrix.png" width="30%"/>

The cells of the matrix are populated by a given user's degree of preference towards an item, which can come in the form of:

1. **explicit feedback:** direct feedback towards an item (e.g., movie ratings which we explored in [Part 1](https://github.com/topspinj/recommender-tutorial/blob/master/part-1-item-item-recommender.ipynb))
2. **implicit feedback:** indirect behaviour towards an item (e.g., purchase history, browsing history, search behaviour)

Implicit feedback makes assumptions about a user's preference based on their actions towards items. Let's take Netflix for example. If you binge-watch a show and blaze through all seasons in a week, there's a high chance that you like that show. However, if you start watching a series and stop halfway through the first episode, there's suspicion to believe that you probably don't like that show. 

<!-- <img src="images/netflix_implicit_feedback.png" width="50%"/> --> 原作者的 Netflix 圖片沒有提供

### Step 1: Import Dependencies

We'll be using the following packages to build our implicit feedback recommender system:

- [numpy](https://numpy.org/)
- [pandas](https://pandas.pydata.org/)
- [implicit](https://github.com/benfred/implicit)
- scipy (specifically, the [csr_matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html) class)

In [2]:
import numpy as np
import pandas as pd
from scipy.sparse import csr_matrix
import implicit
import warnings

warnings.simplefilter(action = "ignore", category = FutureWarning)

### Step 2: Load the Data

Since we're already familiar with MovieLens from Part 1 and 2 of this tutorial series, we'll continue using this dataset. You can access the MovieLens dataset that we'll be working with via this zip file url [here](https://grouplens.org/datasets/movielens/), or directly download [here](http://files.grouplens.org/datasets/movielens/ml-latest-small.zip). We're working with data in `ml-latest-small.zip` and will need to add the following files to our local directory: 
- ratings.csv
- movies.csv

Alternatively, you can access the data here:   
原作者給的網址仍然失效了。
<!-- - https://s3-us-west-2.amazonaws.com/recommender-tutorial/movies.csv
- https://s3-us-west-2.amazonaws.com/recommender-tutorial/ratings.csv -->

In [3]:
ratings = pd.read_csv("data/ratings.csv")
movies = pd.read_csv("data/movies.csv")

In [4]:
ratings.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

For this implicit feedback tutorial, we'll treat movie ratings as the number of times that a user watched a movie. For example, if Jane (a user in our database) gave `Batman` a rating of 1 and `Legally Blonde` a rating of 5, we'll assume that Jane watched Batman one time and Legally Blonde five times. 

### Step 3: Transforming the Data

Similar to [Part 1](https://github.com/topspinj/recommender-tutorial/blob/master/part-1-item-item-recommender.ipynb), we need to transform the `ratings` dataframe into a user-item matrix where rows represent users and columns represent movies. The cells of this matrix will be populated with implicit feedback: in this case, the number of times a user watched a movie. 

The `create_X()` function outputs a sparse matrix **X** with four mapper dictionaries:
- **user_mapper:** maps user id to user index
- **movie_mapper:** maps movie id to movie index
- **user_inv_mapper:** maps user index to user id
- **movie_inv_mapper:** maps movie index to movie id

We need these dictionaries because they map which row and column of the utility matrix corresponds to which user ID and movie ID, respectively.

The **X** (user-item) matrix is a [scipy.sparse.csr_matrix](scipylinkhere) which stores the data sparsely.



<img src="images/user-movie-matrix.png" width="500px" align="left">

In [5]:
def create_X(df):
    """
    Generates a sparse matrix from ratings dataframe.
    
    Args:
        df: pandas dataframe
    
    Returns:
        X: sparse matrix
        user_mapper: dict that maps user id's to user indices
        user_inv_mapper: dict that maps user indices to user id's
        movie_mapper: dict that maps movie id's to movie indices
        movie_inv_mapper: dict that maps movie indices to movie id's
    """
    N = df["userId"].nunique()
    M = df["movieId"].nunique()

    user_mapper = dict(zip(np.unique(df["userId"]), list(range(N))))    
    movie_mapper = dict(zip(np.unique(df["movieId"]), list(range(M))))

    user_inv_mapper = dict(zip(list(range(N)), np.unique(df["userId"])))
    movie_inv_mapper = dict(zip(list(range(M)), np.unique(df["movieId"])))

    user_index = [user_mapper[i] for i in df["userId"]]
    movie_index = [movie_mapper[i] for i in df["movieId"]]

    X = csr_matrix((df["rating"], (movie_index, user_index)), shape = (M, N))

    return X, user_mapper, movie_mapper, user_inv_mapper, movie_inv_mapper

In [6]:
X, user_mapper, movie_mapper, user_inv_mapper, movie_inv_mapper = create_X(ratings)

### Creating Movie Title Mappers

We need to interpret a movie title from its index in the user-item matrix and vice versa. Let's create 2 helper functions that make this interpretation easy:

- `get_movie_index()` - converts a movie title to movie index
    - Note that this function uses [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy)'s string matching to get the approximate movie title match based on the string that gets passed in. This means that you don't need to know the exact spelling and formatting of the title to get the corresponding movie index.
- `get_movie_title()` - converts a movie index to movie title

In [8]:
try:
    import Levenshtein  # noqa: F401 # 這樣才不會有錯誤或警告
except ImportError:
    pass
from fuzzywuzzy import process

def movie_finder(title):
    all_titles = movies["title"].tolist()
    closest_match = process.extractOne(title, all_titles)
    return closest_match[0]

movie_title_mapper = dict(zip(movies["title"], movies["movieId"]))
movie_title_inv_mapper = dict(zip(movies["movieId"], movies["title"]))

def get_movie_index(title):
    fuzzy_title = movie_finder(title)
    movie_id = movie_title_mapper[fuzzy_title]
    movie_idx = movie_mapper[movie_id]
    return movie_idx

def get_movie_title(movie_idx):
    movie_id = movie_inv_mapper[movie_idx]
    title = movie_title_inv_mapper[movie_id]
    return title

It's time to test it out! Let's get the movie index of `Legally Blonde`. 

In [9]:
get_movie_index("Legally Blonde")

3282

Let's pass this index value into `get_movie_title()`. We're expecting Legally Blonde to get returned.

In [10]:
get_movie_title(3282)

'Legally Blonde (2001)'

Great! These helper functions will be useful when we want to interpret our recommender results.

### Step 4: Building Our Implicit Feedback Recommender Model


We've transformed and prepared our data so that we can start creating our recommender model.

The [implicit](https://github.com/benfred/implicit) package is built around a linear algebra technique called [matrix factorization](https://en.wikipedia.org/wiki/Matrix_factorization_(recommender_systems)), which can help us discover latent features underlying the interactions between users and movies. These latent features give a more compact representation of user tastes and item descriptions. **Matrix factorization is particularly useful for very sparse data** and can enhance the quality of recommendations. The algorithm works by factorizing the original user-item matrix into two factor matrices:

- user-factor matrix (n_users, k)
- item-factor matrix (k, n_items)

We are reducing the dimensions of our original matrix into "taste" dimensions. We cannot interpret what each latent feature $k$ represents. However, we could imagine that one latent feature may represent users who like romantic comedies from the 1990s, while another latent feature may represent movies which are independent foreign language films.

$$X_{mn} \approx P_{mk} \times Q_{nk}^T = \hat{X}$$

<img src="images/matrix-factorization.png" width="60%"/>

In traditional matrix factorization, such as SVD, we would attempt to solve the factorization at once which can be very computationally expensive. As a more practical alternative, we can use a technique called `Alternating Least Squares (ALS)` instead. With ALS, we solve for one factor matrix at a time:

- Step 1: hold user-factor matrix fixed and solve for the item-factor matrix
- Step 2: hold item-factor matrix fixed and solve for the user-item matrix

We alternate between Step 1 and 2 above, until the dot product of the item-factor matrix and user-item matrix is approximately equal to the original X (user-item) matrix. This approach is less computationally expensive and can be run in parallel.

The [implicit](https://github.com/benfred/implicit) package implements matrix factorization using Alternating Least Squares (see docs [here](https://implicit.readthedocs.io/en/latest/als.html)，此文件已失效). Let's initiate the model using the `AlternatingLeastSquares` class.

在執行以下程式碼前，先在 Windows（cmd/Anaconda Prompt）執行以下程式，避免開多執行緒搶資源。
- set MKL_NUM_THREADS=1
- set OMP_NUM_THREADS=1
- set OPENBLAS_NUM_THREADS=1

In [11]:
model = implicit.als.AlternatingLeastSquares(factors = 50)

c:\Users\User\anaconda3\envs\recommender_systems\Lib\site-packages\implicit\cpu\als.py:95: RuntimeWarning: Intel MKL BLAS is configured to use 8 threads. It is highly recommended to disable its internal threadpool by setting the environment variable 'MKL_NUM_THREADS=1' or by callng 'threadpoolctl.threadpool_limits(1, "blas")'. Having MKL use a threadpool can lead to severe performance issues
  check_blas_config()


This model comes with a couple of hyperparameters that can be tuned to generate optimal results:

- factors ($k$): number of latent factors,
- regularization ($\lambda$): prevents the model from overfitting during training

In this tutorial, we'll set $k = 50$ and $\lambda = 0.01$ (the default). In a real-world scenario, I highly recommend tuning these hyperparameters before generating recommendations to generate optimal results.

The next step is to fit our model with our user-item matrix. 

In [13]:
model.fit(X) # 訓練時，模型認定列為 user、欄為 item

  0%|          | 0/15 [00:00<?, ?it/s]

Now, let's test out the model's recommendations. We can use the model's `similar_items()` method which returns the most relevant movies of a given movie. We can use our helpful `get_movie_index()` function to get the movie index of the movie that we're interested in.


similar_items(itemid, N = 10)
- 輸入：一個 item（例如某部電影的 ID）。
- 輸出：和這個 item 最相似的 N 個其他 items。
- 用途：做 `item-based` 推薦，例如：
-「喜歡《The Dark Knight》的觀眾，可能也會喜歡這些電影」。
- 背後依據：用訓練後的 item latent factors 計算 cosine similarity 或內積。

In [25]:
movie_of_interest = "forrest gump"

movie_index = get_movie_index(movie_of_interest)
related = model.similar_items(movie_index) # 模型的輸入是 csr 的 movie_index
# 執行緒（環境變數）、implicit 版本、ALS 參數都有可能影響結果
related # 第一個內容是自己本身，此外，我的結果和原作者不一樣，形式也有改變

(array([314, 187, 284, 126, 167,  28,  26, 308, 516, 553], dtype=int32),
 array([1.0000001 , 0.6051691 , 0.5822848 , 0.5524658 , 0.5268991 ,
        0.5170927 , 0.5002023 , 0.4826193 , 0.437059  , 0.40624124],
       dtype=float32))

In [11]:
movie_of_interest = 'forrest gump'

movie_index = get_movie_index(movie_of_interest)
related = model.similar_items(movie_index)
related

[(314, 0.9999999),
 (277, 0.87162775),
 (510, 0.8275397),
 (257, 0.82705915),
 (97, 0.7511661),
 (461, 0.7223397),
 (418, 0.6889433),
 (1938, 0.6659612),
 (123, 0.6441393),
 (43, 0.61916924)]

The output of `similar_items()` is not user-friendly. We'll need to use our `get_movie_title()` function to interpret what our results are. 

In [39]:
# 因為 related 形式和作者不一樣，所以處理方式也會不一樣
print(f"Because you watched {movie_finder(movie_of_interest)}...")

for r in related[0]:
    # print(r)
    recommended_title = get_movie_title(r)
    if recommended_title != movie_finder(movie_of_interest): # 去除掉自己本身，所以會剩下 9 部電影
        print(recommended_title)

Because you watched Forrest Gump (1994)...
Cure, The (1995)
To Live (Huozhe) (1994)
Batman Forever (1995)
Strange Days (1995)
City of Lost Children, The (Cité des enfants perdus, La) (1995)
Now and Then (1995)
Client, The (1994)
Love and a .45 (1994)
Kids in the Hall: Brain Candy (1996)


In [12]:
print(f"Because you watched {movie_finder(movie_of_interest)}...")
for r in related:
    recommended_title = get_movie_title(r[0])
    if recommended_title != movie_finder(movie_of_interest):
        print(recommended_title)

Because you watched Forrest Gump (1994)...
Shawshank Redemption, The (1994)
Silence of the Lambs, The (1991)
Pulp Fiction (1994)
Braveheart (1995)
Schindler's List (1993)
Jurassic Park (1993)
Matrix, The (1999)
Apollo 13 (1995)
Seven (a.k.a. Se7en) (1995)


When we treat user ratings as implicit feedback, the results look pretty good! You can test out other movies by changing the `movie_of_interest` variable.

### Step 5: Generating User-Item Recommendations

A cool feature of [implicit](https://github.com/benfred/implicit) is that you can pull personalized recommendations for a given user. Let's test it out on a user in our dataset.

In [14]:
user_id = 95

In [15]:
user_ratings = ratings[ratings["userId"] == user_id].merge(movies[["movieId", "title"]])
user_ratings = user_ratings.sort_values("rating", ascending = False)
print(f"Number of movies rated (watched here) by user {user_id}: {user_ratings["movieId"].nunique()}")

Number of movies rated (watched here) by user 95: 168


User 95 watched 168 movies. Their highest rated movies are below:

In [49]:
top_5 = user_ratings.head() # 排序和作者不一樣
top_5

userId  movieId  rating   timestamp  \
3       95       29     5.0  1043340522   
2       95       25     5.0  1044744590   
22      95     1036     5.0  1043339248   
20      95      924     5.0  1043340392   
18      95      741     5.0  1043340485   

                                                title  
3   City of Lost Children, The (Cité des enfants p...  
2                            Leaving Las Vegas (1995)  
22                                    Die Hard (1988)  
20                       2001: A Space Odyssey (1968)  
18         Ghost in the Shell (Kôkaku kidôtai) (1995)

In [15]:
user_ratings = ratings[ratings['userId']==user_id].merge(movies[['movieId', 'title']])
user_ratings = user_ratings.sort_values('rating', ascending=False)
top_5 = user_ratings.head()
top_5

userId  movieId  rating   timestamp  \
24      95     1089     5.0  1048382826   
34      95     1221     5.0  1043340018   
83      95     3019     5.0  1043340112   
26      95     1175     5.0  1105400882   
27      95     1196     5.0  1043340018   

                                                title  
24                              Reservoir Dogs (1992)  
34                     Godfather: Part II, The (1974)  
83                            Drugstore Cowboy (1989)  
26                                Delicatessen (1991)  
27  Star Wars: Episode V - The Empire Strikes Back...

Their lowest rated movies:

In [50]:
bottom_5 = user_ratings.tail() # 1 分的和作者一樣，2 分的有些不一樣
bottom_5

userId  movieId  rating   timestamp                                 title
82       95     2990     2.0  1043339283                Licence to Kill (1989)
43       95     1381     2.0  1043339908                       Grease 2 (1982)
122      95     5283     2.0  1043339957  National Lampoon's Van Wilder (2002)
109      95     4732     1.0  1043339283                     Bubble Boy (2001)
164      95     7373     1.0  1105401093                        Hellboy (2004)

In [16]:
bottom_5 = user_ratings[user_ratings['rating']<3].tail()
bottom_5

userId  movieId  rating   timestamp                                 title
93       95     3690     2.0  1043339908                Porky's Revenge (1985)
122      95     5283     2.0  1043339957  National Lampoon's Van Wilder (2002)
100      95     4015     2.0  1043339957          Dude, Where's My Car? (2000)
164      95     7373     1.0  1105401093                        Hellboy (2004)
109      95     4732     1.0  1043339283                     Bubble Boy (2001)

Based on their preferences above, we can get a sense that user 95 likes action and crime movies from the early 1990's over light-hearted American comedies from the early 2000's. Let's see what recommendations our model will generate for user 95.

We'll use the `recommend()` method, which takes in the user index of interest and transposed user-item matrix. 

recommend(userid, user_items, N = 10, ...)
- 輸入：一個 user ID + 該使用者的互動紀錄（user_items 稀疏矩陣）。
- 輸出：推薦給這個 user 的 N 個 items。
- 用途：做 `user-based` 推薦，例如：
-「使用者 95 可能喜歡這些電影」。
- 背後依據：用訓練後的 user latent factors 和 item latent factors 算出偏好分數，再排序。

In [16]:
print("X_t:", X.T.tocsr().shape) # 預期 (users = 610, items = 9724)
print("model.user_factors.shape:", model.user_factors.shape)
print("model.item_factors.shape:", model.item_factors.shape)

X_t: (610, 9724)
model.user_factors.shape: (9724, 50)
model.item_factors.shape: (610, 50)


In [27]:
X_t = X.T.tocsr().astype(np.float32) # X: user x item
model = implicit.als.AlternatingLeastSquares(factors = 50) # ALS/BPR(Bayesian Personalized Rankin)等
model.fit(X_t)

  0%|          | 0/15 [00:00<?, ?it/s]

In [28]:
print("X_t:", X_t.shape) # 期待 (users = 610, items = 9724)
print("model.user_factors.shape:", model.user_factors.shape)
print("model.item_factors.shape:", model.item_factors.shape)

X_t: (610, 9724)
model.user_factors.shape: (610, 50)
model.item_factors.shape: (9724, 50)


In [29]:
print(X_t.shape)

user_idx = user_mapper[user_id]
print(user_idx)

user_row = X_t[user_idx]
print(user_row.shape)

recommendations = model.recommend(user_idx, user_row) # 第二個參數只需要傳入一列
recommendations

(610, 9724)
94
(1, 9724)


(array([1210,  855,  922, 3633,  915,  898, 4155,    5,   15,  828],
       dtype=int32),
 array([1.0979283 , 1.0691261 , 0.9706177 , 0.9677502 , 0.90129566,
        0.89468205, 0.8867651 , 0.8860675 , 0.82140845, 0.778286  ],
       dtype=float32))

In [ ]:
# 原作者的作法無法使用了
# X_t = X.T.tocsr()

# user_idx = user_mapper[user_id]
# recommendations = model.recommend(user_idx, X_t)
# recommendations

[(855, 1.127779),
 (1043, 0.98673713),
 (1210, 0.9256185),
 (3633, 0.90900886),
 (1978, 0.8929481),
 (4155, 0.84075284),
 (2979, 0.82858247),
 (3609, 0.78015),
 (4791, 0.7672245),
 (4010, 0.7530525)]

<img src=images/tocsr.png width=50% height=50%>

We can't interpret the results as is since movies are represented by their index. We'll have to loop over the list of recommendations and get the movie title for each movie index. 

In [32]:
for r in recommendations[0]:
    # print(r)
    recommended_title = get_movie_title(r)
    print(recommended_title)

Hunt for Red October, The (1990)
Abyss, The (1989)
Full Metal Jacket (1987)
Lord of the Rings: The Fellowship of the Ring, The (2001)
Army of Darkness (1993)
Princess Bride, The (1987)
Chicago (2002)
Heat (1995)
Casino (1995)
Platoon (1986)


In [ ]:
# 作者的做法已經不可行了
# for r in recommendations:
#     recommended_title = get_movie_title(r[0])
#     print(recommended_title)

Abyss, The (1989)
Star Trek: First Contact (1996)
Hunt for Red October, The (1990)
Lord of the Rings: The Fellowship of the Ring, The (2001)
Star Wars: Episode I - The Phantom Menace (1999)
Chicago (2002)
Crouching Tiger, Hidden Dragon (Wo hu cang long) (2000)
Ocean's Eleven (2001)
Lord of the Rings: The Return of the King, The (2003)
Punch-Drunk Love (2002)


User 95's recommendations consist of action, crime, and thrillers. None of their recommendations are comedies. 

# Extra：附上電影的風格以驗證準確性

In [62]:
# 每次執行的結果會不一樣
# 原作者說 user 95 likes action and crime movies from the early 1990's over light-hearted American comedies from the early 2000's.
# 我看感覺也不一定
def implicit_feedback_recommendations(user_id):
    user_index = user_mapper[user_id]
    user_row = X_t[user_index]
    model = implicit.als.AlternatingLeastSquares(factors = 50)
    model.fit(X_t)
    recommendations = model.recommend(user_index, user_row)
    movie_indices = recommendations[0]
    scores = recommendations[1]
    print("推薦給使用者 ID 為", user_id)
    for index in range(len(movie_indices)):
        movie_title = get_movie_title(movie_indices[index])
        # print(movie_title)
        movie_id = movie_inv_mapper[movie_indices[index]]
        # print(movie_indices[index], movie_id)
        score = round(scores[index], 2)
        print(f"電影名稱: {movie_title}, 風格: {movies.iloc[movie_id]["genres"]}, 相關程度: {score}")

implicit_feedback_recommendations(user_id)

  0%|          | 0/15 [00:00<?, ?it/s]

推薦給使用者 ID 為 95
電影名稱: Lord of the Rings: The Fellowship of the Ring, The (2001), 風格: Comedy|Musical, 相關程度: 1.0399999618530273
電影名稱: Princess Bride, The (1987), 風格: Comedy|Drama, 相關程度: 0.9900000095367432
電影名稱: Abyss, The (1989), 風格: Children|Comedy, 相關程度: 0.9800000190734863
電影名稱: Casino (1995), 風格: Drama|Romance, 相關程度: 0.9100000262260437
電影名稱: Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb (1964), 風格: Comedy|Drama, 相關程度: 0.8999999761581421
電影名稱: Hunt for Red October, The (1990), 風格: Children|Comedy, 相關程度: 0.8999999761581421
電影名稱: Crouching Tiger, Hidden Dragon (Wo hu cang long) (2000), 風格: Drama, 相關程度: 0.8899999856948853
電影名稱: Blues Brothers, The (1980), 風格: Crime|Drama|Mystery|Thriller, 相關程度: 0.8899999856948853
電影名稱: Lord of the Rings: The Return of the King, The (2003), 風格: Action|Sci-Fi|Thriller, 相關程度: 0.8500000238418579
電影名稱: Star Wars: Episode I - The Phantom Menace (1999), 風格: Drama, 相關程度: 0.8399999737739563
